### DISCUSSION TITLE
Is it possible to call a zustand setState directely inside a react component without using useEffect ?

### DISCUSSION THREAD

**Discussion by omarnemer** [pmndrs/zustand#1831] [Q&A]
Hi, maybe the title is confusing, but I'll try to explain. I'm using react query v5 to fetch server data. When the data is fetched, I want to set a zustand state called "total" with the length of that data. We need that length in the parent component. I was doing the same process with React query OnSuccess callback without zustand, but onSuccess has been removed in v5. Now, when the data arrive in that component, I don't want to use an useEffect to set that total state. My question is, can we just do at the bottom of the component before rendering:
`function DayView(){`
`  const setTotal = useStore(
    (state) =>  state.setTotal,
    shallow
   )`
`  const query = useQuery(...)`
`  if(query.data)`
`   seTotal(data.length)`
`}`
We are not using useEffect. If we do that with normal a state, it will create an infinite loop, but with a zustand state, it seems to work and does not cause an infinite loop. This does not seem right.

Thank you for your help, I hope it's clear.

**Comment by dai-shi** [comment]
It should be fine, if React doesn't warn.

**Reply by omarnemer** [reply]
Hi @dai-shi , okay i'll try to do it this way, thank you for your response.

**Comment by dbritto-dev** [comment]
@omarnemer I have exp using react-query I'll create a little example for you

**Reply by omarnemer** [reply]
hi @dbritto-dev, thanks for your response, It would be very helpful, thank you.
